File: widgetssdk/src/main/java/com/glia/widgets/core/dialog/domain/IsShowEnableCallNotificationChannelDialogUseCase.java
package com.glia.widgets.core.dialog.domain;

import com.glia.widgets.core.dialog.PermissionDialogManager;
import com.glia.widgets.core.permissions.PermissionManager;

public class IsShowEnableCallNotificationChannelDialogUseCase {
    private final PermissionDialogManager permissionDialogManager;
    private final PermissionManager permissionManager;

    public IsShowEnableCallNotificationChannelDialogUseCase(
            PermissionManager permissionManager,
            PermissionDialogManager permissionDialogManager
    ) {
        this.permissionManager = permissionManager;
        this.permissionDialogManager = permissionDialogManager;
    }

    public boolean execute() {
        return isCallNotificationChannelNotEnabled() && hasNotShownCallNotificationNotEnabledRequest();
    }

    private boolean isCallNotificationChannelNotEnabled() {
        return !this.permissionManager.hasCallNotificationChannelEnabled();
    }

    private boolean hasNotShownCallNotificationNotEnabledRequest() {
        return !this.permissionDialogManager.hasEnableCallNotificationChannelRequestShown();
    }
}

File: widgetssdk/src/main/java/com/glia/widgets/core/dialog/domain/IsShowEnableCallNotificationChannelDialogUseCase.kt
package com.glia.widgets.core.dialog.domain

import com.glia.widgets.core.dialog.PermissionDialogManager
import com.glia.widgets.core.permissions.domain.HasCallNotificationChannelEnabledUseCase

internal interface IsShowEnableCallNotificationChannelDialogUseCase {
    operator fun invoke(): Boolean
}

internal class IsShowEnableCallNotificationChannelDialogUseCaseImpl(
    private val hasCallNotificationChannelEnabledUseCase: HasCallNotificationChannelEnabledUseCase,
    private val permissionDialogManager: PermissionDialogManager
) : IsShowEnableCallNotificationChannelDialogUseCase {
    private val isCallNotificationChannelNotEnabled: Boolean get() = !hasCallNotificationChannelEnabledUseCase()
    private val hasNotShownCallNotificationNotEnabledRequest: Boolean get() = !permissionDialogManager.hasEnableCallNotificationChannelRequestShown()

    override fun invoke(): Boolean = isCallNotificationChannelNotEnabled && hasNotShownCallNotificationNotEnabledRequest
}
